### ISSUE TITLE
Interactive REPL Interface

### ISSUE THREAD

**Issue by nixpulvis** [nixpulvis/oursh#5]
In addition to the basic support for reading programs from a non-interactive input, we need to have the ability to run in interactive mode, with prompts, and the full suite of features.

A crazy idea could be to allow interactive mode for non-interactive scripts, where coded user input like `gi<up-arrow>` would run the last `gi` prefixed command.

Key features:

- [ ] Prompt
- [ ] Cursor movement
- [ ] History
- [ ] Completion
- [ ] Syntax highlighting
- [ ] Multi-line programs (`\` line continuations, and incomplete syntax)
- [ ] Interactive signals?

**Comment by nixpulvis** [comment]
I want to move all stream processing out of `main.rs` and into the fancy new `repl` module. I believe [`BufReadDecoder`](https://docs.rs/utf-8/0.7.4/utf8/struct.BufReadDecoder.html) will be needed.

**Comment by nixpulvis** [comment]
The initial `sh` style completion feature will be simple and fast enough to run inline, but after that point a more robust asynchronous implementation is needed.

**Comment by dread-uo** [comment]
What would the mouse features be? Asking just out of curiosity :p
By the way, nice project!

**Comment by nixpulvis** [comment]
@dread-uo I've been so busy writing all this POSIX stuff I haven't thought much about it. I spent a good part of tonight working on making the raw terminal mode a cargo feature... mouse actions could go on top of that.

Looking at [`MouseEvent`](https://docs.rs/termion/1.5.1/termion/event/enum.MouseEvent.html) looks like we get everything we'd need. Terminals are netoriously known for not being heavy on mouse interactions, but I think there are a few *nice* things we could try. As a general rule of thumb, mouse actions should be things you are doing **not** as part of a normal text REPL workflow. For example a mouse button to close a search mode would make me cry.

1. I already have dreams of a mode that operates like a shared tmux session, perhaps the mouse could show up as a colored cursor on the other persons terminal window.
2. While many terminal emulators support clicking links and things like that, we could have our own logic, and include more kinds of links.
3. There's something really interesting about having shell ASTs and the ability to click the syntax, maybe something like https://explainshell.com/ could be made to work well.

Basically who knows at this point. Thanks for the question. Let me know if you end up using the shell at all, or want to help :wink:

**Comment by nixpulvis** [comment]
In order to support current and future needs for completion, here is a more complete design.

- `complete_command`
  - `complete_builtin`
  - `complete_function`
  - `complete_executable`
- `complete_path` (includes `~`, and globs)
- `complete_variable`
- `complete_job`
- `complete_syntax`

```rs
enum Completion {
    None,
    Partial(String, HashSet<String>),
    Complete(String),
}
```

Still need to figure out the best place to put the completion for the fish style last history completion.

**Comment by nixpulvis** [comment]
A good list of features to support to be inline with `sh` can be found [here](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/sh.html) under the section "Command Line Editing".
